File: VB/WinSolution3.Module/Updater.vb
Imports Microsoft.VisualBasic
Imports System

Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Updating
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.ExpressApp.Reports
Imports System.Collections

Namespace WinSolution3.Module
	Public Class Updater
		Inherits ModuleUpdater
		Public Sub New(ByVal objectSpace As IObjectSpace, ByVal currentDBVersion As Version)
			MyBase.New(objectSpace, currentDBVersion)
		End Sub
		Public Overrides Sub UpdateDatabaseAfterUpdateSchema()
			MyBase.UpdateDatabaseAfterUpdateSchema()
			CreateReport("DetailReport")
			CreateReport("MasterReport")
			CreateMaster("Master 1", New Object(){ CreateDetail("Detail 1"), CreateDetail("Detail 2"), CreateDetail("Detail 3")})
			CreateMaster("Master 2", New Object(){ CreateDetail("Detail 4"), CreateDetail("Detail 5")})
			CreateMaster("Master 3", New Object(){ CreateDetail("Detail 6"), CreateDetail("Detail 7"), CreateDetail("Detail 8"), CreateDetail("Detail 9")})
		End Sub
		Private Sub CreateReport(ByVal reportName As String)
			Dim reportdata As ReportData = ObjectSpace.FindObject(Of ReportData)(New BinaryOperator("Name", reportName))
			If reportdata Is Nothing Then
				reportdata = ObjectSpace.CreateObject(Of ReportData)()
				Dim rep As New XafReport()
				rep.ObjectSpace = ObjectSpace
				rep.LoadLayout(Me.GetType().Assembly.GetManifestResourceStream("SavedReports." & reportName & ".repx"))
				rep.ReportName = reportName
				reportdata.SaveReport(rep)
				reportdata.Save()
			End If
		End Sub
		Private Function CreateMaster(ByVal name As String, ByVal details As ICollection) As DomainObject1
			Dim master As DomainObject1 = ObjectSpace.FindObject(Of DomainObject1)(New BinaryOperator("Name", name))
			If master Is Nothing Then
				master = ObjectSpace.CreateObject(Of DomainObject1)()
				master.Name = name
				For Each detail As DomainObject2 In details
					master.DomainObject2s.Add(detail)
				Next detail
				master.Save()
			End If
			Return master
		End Function
		Private Function CreateDetail(ByVal name As String) As DomainObject2
			Dim detail As DomainObject2 = ObjectSpace.FindObject(Of DomainObject2)(New BinaryOperator("Name", name))
			If detail Is Nothing Then
				detail = ObjectSpace.CreateObject(Of DomainObject2)()
				detail.Name = name
				detail.Save()
			End If
			Return detail
		End Function
	End Class
End Namespace

File: CS/WinSolution3.Module/Updater.cs
using System;

using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Updating;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;
using DevExpress.Persistent.BaseImpl;
using DevExpress.ExpressApp.Reports;
using System.Collections;

namespace WinSolution3.Module {
    public class Updater : ModuleUpdater {
        public Updater(IObjectSpace objectSpace, Version currentDBVersion) : base(objectSpace, currentDBVersion) { }
        public override void UpdateDatabaseAfterUpdateSchema() {
            base.UpdateDatabaseAfterUpdateSchema();
            CreateReport("DetailReport");
            CreateReport("MasterReport");
            CreateMaster("Master 1", new object[]{  CreateDetail("Detail 1"),
                                                    CreateDetail("Detail 2"),
                                                    CreateDetail("Detail 3")});
            CreateMaster("Master 2", new object[]{  CreateDetail("Detail 4"),
                                                    CreateDetail("Detail 5")});
            CreateMaster("Master 3", new object[]{  CreateDetail("Detail 6"),
                                                    CreateDetail("Detail 7"),
                                                    CreateDetail("Detail 8"),
                                                    CreateDetail("Detail 9")});
        }
        private void CreateReport(string reportName) {
            ReportData reportdata = ObjectSpace.FindObject<ReportData>(new BinaryOperator("Name", reportName));
            if (reportdata == null) {
                reportdata = ObjectSpace.CreateObject<ReportData>();
                XafReport rep = new XafReport();
                rep.ObjectSpace = ObjectSpace;
                rep.LoadLayout(GetType().Assembly.GetManifestResourceStream(
                   "WinSolution3.Module.SavedReports." + reportName + ".repx"));
                rep.ReportName = reportName;
                reportdata.SaveReport(rep);
                reportdata.Save();
            }
        }
        private DomainObject1 CreateMaster(string name, ICollection details) {
            DomainObject1 master = ObjectSpace.FindObject<DomainObject1>(new BinaryOperator("Name", name));
            if (master == null) {
                master = ObjectSpace.CreateObject<DomainObject1>();
                master.Name = name;
                foreach(DomainObject2 detail in details) {
                    master.DomainObject2s.Add(detail);
                }
                master.Save();
            }
            return master;
        }
        private DomainObject2 CreateDetail(string name) {
            DomainObject2 detail = ObjectSpace.FindObject<DomainObject2>(new BinaryOperator("Name", name));
            if (detail == null) {
                detail = ObjectSpace.CreateObject<DomainObject2>();
                detail.Name = name;
                detail.Save();
            }
            return detail;
        }
    }
}
